flink-yarn/src/main/java/org/apache/flink/yarn/YarnClusterDescriptor.java

```diff
@@ -1588,16 +1589,27 @@ private void addPluginsFoldersToShipFiles(Collection<File> effectiveShipFiles) {
 		}
 	}
 
-	ContainerLaunchContext setupApplicationMasterContainer(
+	protected ContainerLaunchContext setupApplicationMasterContainer(
+			String appId,
 			String yarnClusterEntrypoint,
 			boolean hasLogback,
 			boolean hasLog4j,
 			boolean hasKrb5,
 			int jobManagerMemoryMb) {
 		// ------------------ Prepare Application Master Container  ------------------------------
 
+		String javaOpts = "";
+		// Add default gc logging options if enabled
+		javaOpts += BootstrapTools.getGCLoggingOpts(ApplicationConstants.LOG_DIR_EXPANSION_VAR, flinkConfiguration);
+		// Add default heap dump options if enabled
+		javaOpts += " " + BootstrapTools.getHeapdumpOpts(
+			appId,
+			"jobmanager",
+			ApplicationConstants.LOG_DIR_EXPANSION_VAR,
+			flinkConfiguration);
 		// respect custom JVM options in the YAML file
-		String javaOpts = flinkConfiguration.getString(CoreOptions.FLINK_JVM_OPTIONS);
+		javaOpts += " " + flinkConfiguration.getString(CoreOptions.FLINK_JVM_OPTIONS);
```

Comment:
This looks like duplicate code which we also have in `BootstrapTools.getTaskManagerShellCommand`. Please deduplicate.

Reply:
Done